sw/device/lib/testing/usb_testutils_streams.c

```diff
@@ -275,6 +437,35 @@ static status_t rx_show(void *stream_v, dif_usbdev_rx_packet_info_t packet_info,
   return OK_STATUS();
 }
 
+// Set the count of already-initialized streams, and apportion the available
+// tx buffers among the streams.
+bool usb_testutils_streams_count_set(usb_testutils_streams_ctx_t *ctx,
+                                     unsigned nstreams) {
+  // Too many streams?
+  if (nstreams > USBUTILS_STREAMS_MAX) {
+    return false;
+  }
+
+  // Decide how many buffers each endpoint may queue up for transmission;
+  // we must ensure that there are buffers available for reception, and we
+  // do not want any endpoint to starve another
+  for (unsigned s = 0U; s < nstreams; s++) {
+    // This is slightly overspending the available buffers, leaving the
+    //   endpoints to vie for the final few buffers, so it's important that
+    //   we limit the total number of buffers across all endpoints too
+    unsigned ep = ctx->streams[s].tx_ep;
+    ctx->tx_bufs_queued[ep] = 0U;
+    ctx->tx_bufs_limit[ep] =
+        (USBUTILS_STREAMS_TXBUF_MAX + nstreams - 1) / nstreams;
```

Comment:
Does this oversubscribing of buffers enter the potential for random failures? (No analysis done on my part; this checking was merely prodded by the comments)

Reply:
No, the code in usb_testutils_stream_service() will only attempt to claim a 'send buffer' if the budgets would not be exceeded, and will then only proceed if it successfully claims a buffer from the DIF. Provided that all streams are regularly serviced it balances out.

`  if (tx_more &&                              // More bytes to transfer?
      nqueued < ctx->tx_bufs_limit[tx_ep] &&  // Endpoint allowed buffer?
      ctx->tx_queued_total <
          USBUTILS_STREAMS_TXBUF_MAX) {  // Total buffers not exceeded?`

Reply:
Thanks!